crates/handlers/src/oauth2/introspection.rs

```diff
@@ -205,28 +222,41 @@ pub(crate) async fn post(
     mut repo: BoxRepository,
     activity_tracker: ActivityTracker,
     State(encrypter): State<Encrypter>,
+    State(homeserver): State<Arc<dyn HomeserverConnection>>,
     headers: HeaderMap,
-    client_authorization: ClientAuthorization<IntrospectionRequest>,
+    ClientAuthorization { credentials, form }: ClientAuthorization<IntrospectionRequest>,
 ) -> Result<impl IntoResponse, RouteError> {
-    let client = client_authorization
-        .credentials
-        .fetch(&mut repo)
-        .await?
-        .ok_or(RouteError::ClientNotFound)?;
-
-    let method = match &client.token_endpoint_auth_method {
-        None | Some(OAuthClientAuthenticationMethod::None) => {
-            return Err(RouteError::NotAllowed(client.id));
+    if let Some(token) = credentials.bearer_token() {
+        // If the client presented a bearer token, we check with the homeserver
+        // connection if it is allowed to use the introspection endpoint
```

Comment:
```suggestion
        // configuration if it is allowed to use the introspection endpoint
```
I think makes a bit more sense?
We aren't actually making a request to the HS to validate this token. (Maybe it shouldn't be async)

Reply:
it's just that this trait is called `HomeserverConnection`, but no strong opinion here

I made it async thinking that maybe we may want to do something else for verifying the token but maybe I'm optimising a future change too early 🤷